### PR TITLE
[Core] About MessageFilter

### LINKED ISSUE #548: [Bug?]: 发送图片时会产生一张url不正确的空图片
### 请确保您已阅读以上注意事项，并勾选下方的确认框。

- [X] 我已经仔细阅读上述内容
- [X] 我已经使用 [最新构建](https://github.com/LagrangeDev/Lagrange.Core/actions/) 测试过，问题依旧存在。
- [X] 我已经在 [Issue Tracker](https://github.com/LagrangeDev/Lagrange.Core/issues) 中找过我要提出的问题，没有找到相同问题的ISSUE。
- [X] 我已知晓并同意，此处仅用于汇报程序中存在的问题。若这个 Issue 是关于其他非程序本身问题，则我的 Issue 可能会被无条件自动关闭或/并锁定。（其它疑问请考虑加入TG群询问或在discussions中提问）

### Lagrange项目

OneBot

### 所使用/依赖的Lagrange项目对应的commit

1a8f1710

### 运行环境

MacOS

### 运行架构

arm64

### 连接方式

正向 WebSocket

### 重现步骤

如题

### 期望的结果是什么？

一张包含正确url的图片

### 实际的结果是什么？

两张图片，第一张是旧的url格式并且404

### 简单的复现代码/链接（可选）

_No response_

### Trace 级别日志记录（可选）

```Shell
Lagrange.OneBot Version: 0.0.3+1a8f1710ad3d205d56e22661c32449c7a2f4eefc

warn: Lagrange.OneBot.Utility.MusicSigner[0]
      MusicSignServer is not available, sign may be failed
info: Lagrange.OneBot.LagrangeAppBuilder[0]
      Indexing in the database...
info: Lagrange.OneBot.LagrangeAppBuilder[0]
      The first indexing of the old database will load the entire database into memory.
info: Lagrange.OneBot.LagrangeAppBuilder[0]
      If this is the first time creating an index for the old database, please restart the application
info: Microsoft.Hosting.Lifetime[0]
      Application started. Press Ctrl+C to shut down.
info: Microsoft.Hosting.Lifetime[0]
      Hosting environment: Production
info: Microsoft.Hosting.Lifetime[0]
      Content root path: /Users/u/lgr/Lagrange.Core/Lagrange.OneBot/bin/Debug/net8.0
info: Lagrange.OneBot.LagrangeApp[0]
      Lagrange.OneBot Implementation has started
info: Lagrange.OneBot.LagrangeApp[0]
      Protocol:  | 3.2.10-25765
info: Lagrange.Core.BotContext[0]
      [2024-08-24 06:17:17] [SocketContext] [VERBOSE]: Server: http://39.156.126.229:8080/ Latency: 30
info: Lagrange.Core.BotContext[0]
      [2024-08-24 06:17:17] [SocketContext] [VERBOSE]: Server: http://111.30.187.245:8080/ Latency: 27
info: Lagrange.Core.BotContext[0]
      [2024-08-24 06:17:17] [SocketContext] [VERBOSE]: Server: http://111.30.187.201:8080/ Latency: 29
info: Lagrange.Core.BotContext[0]
      [2024-08-24 06:17:17] [WtExchangeLogic] [INFORMATION]: Session has not expired, using session to login and register status
trce: Lagrange.Core.BotContext[0]
      [2024-08-24 06:17:17] [ServiceContext] [DEBUG]: Outgoing SSOFrame: trpc.qq_new_tech.status_svc.StatusService.Register
trce: Lagrange.Core.BotContext[0]
      [2024-08-24 06:17:17] [ServiceContext] [DEBUG]: Incoming SSOFrame: trpc.qq_new_tech.status_svc.StatusService.Register
info: Lagrange.Core.BotContext[0]
      [2024-08-24 06:17:17] [WtExchangeLogic] [INFORMATION]: Register Status: register success
info: Lagrange.Core.BotContext[0]
      [2024-08-24 06:17:17] [WtExchangeLogic] [INFORMATION]: AutoReLogin Enabled, session would be refreshed in 15 days period
info: Lagrange.OneBot.LagrangeApp[0]
      Bot Online: ***
info: Lagrange.OneBot.Core.Network.LagrangeWebSvcCollection[1]
      [LagrangeWebSvcCollection]: Multi Connection has been configured
info: Lagrange.OneBot.Core.Network.Service.ForwardWSService[10]
      The server is started at http://127.0.0.1:6700/
info: Lagrange.OneBot.Core.Network.Service.ForwardWSService[11]
      Connect(72a874e5-6ea5-4e38-978e-866ca7222a93)
trce: Lagrange.OneBot.Core.Network.Service.ForwardWSService[13]
      Send(72a874e5-6ea5-4e38-978e-866ca7222a93) {"sub_type":"connect","meta_event_type":"lifecycle","time":1724451437,"self_id":***,"post_type":"meta_event"}
trce: Lagrange.OneBot.Core.Network.Service.ForwardWSService[13]
      Send(72a874e5-6ea5-4e38-978e-866ca7222a93) {"interval":5000,"status":{"app_initialized":true,"app_enabled":true,"app_good":true,"online":true,"good":true},"meta_event_type":"heartbeat","time":1724451437,"self_id":***,"post_type":"meta_event"}
trce: Lagrange.OneBot.Core.Network.Service.ForwardWSService[13]
      Send(72a874e5-6ea5-4e38-978e-866ca7222a93) {"interval":5000,"status":{"app_initialized":true,"app_enabled":true,"app_good":true,"online":true,"good":true},"meta_event_type":"heartbeat","time":1724451442,"self_id":***,"post_type":"meta_event"}
trce: Lagrange.Core.BotContext[0]
      [2024-08-24 06:17:24] [ServiceContext] [DEBUG]: Incoming SSOFrame: trpc.msg.olpush.OlPushService.MsgPush
trce: Lagrange.Core.BotContext[0]
      [2024-08-24 06:17:24] [ServiceContext] [DEBUG]: Outgoing SSOFrame: OidbSvcTrpcTcp.0xfe7_3
trce: Lagrange.Core.BotContext[0]
      [2024-08-24 06:17:25] [ServiceContext] [DEBUG]: Incoming SSOFrame: OidbSvcTrpcTcp.0xfe7_3
info: Lagrange.Core.BotContext[0]
      [2024-08-24 06:17:25] [MessagingLogic] [VERBOSE]: [MessageChain(***)(***)] [Mention]: @rika(***) | [Text]:  #usage rss
trce: Lagrange.OneBot.Core.Network.Service.ForwardWSService[13]
      Send(72a874e5-6ea5-4e38-978e-866ca7222a93) {"message_type":"group","sub_type":"normal","message_id":291336708,"group_id":***,"user_id":***,"anonymous":null,"message":[{"type":"at","data":{"qq":"***","name":"@rika"}},{"type":"text","data":{"text":" #usage rss"}}],"raw_message":"[CQ:at,qq=***,name=@rika] #usage rss","font":0,"sender":{"user_id":***,"nickname":"","card":"","sex":"unknown","age":0,"area":"","level":"43","role":"owner","title":""},"time":1724451445,"self_id":***,"post_type":"message"}
trce: Lagrange.OneBot.Core.Network.Service.ForwardWSService[12]
      Receive(72a874e5-6ea5-4e38-978e-866ca7222a93) {"action":"send_group_msg","params":{"group_id":***,"message":[{"type":"image","data":{"file":"base64:///9j/2wCEAAoHBwgHBgoICAgLCgoLDhgQDg0NDh0VFhEYIx8lJCIfIiEmKzcvJik0KSEiMEExNDk7Pj4+JS5ESUM8SDc9PjsBCgsLDg0OHBAQHDsoIig7Ozs7Ozs7Ozs7Ozs7Ozs7Ozs7Ozs7Ozs7Ozs7Ozs7Ozs7Ozs7Ozs7Ozs7Ozs7Ozs7O//AABEIA94E+AMBIgACEQEDEQH/xAGiAAABBQEBAQEBAQAAAAAAAAAAAQIDBAUGBwgJCgsQAAIBAwMCBAMFBQQEAAABfQECAwAEEQUSITFBBhNRYQcicRQygZGhCCNCscEVUtHwJDNicoIJChYXGBkaJSYnKCkqNDU2Nzg5OkNERUZHSElKU1RVVldYWVpjZGVmZ2hpanN0dXZ3eHl6g4SFhoeIiYqSk5SVlpeYmZqio6Slpqeoqaqys7S1tre4ubrCw8TFxsfIycrS09TV1tfY2drh4uPk5ebn6Onq8fLz9PX29/j5+gEAAwEBAQEBAQEBAQAAAAAAAAECAwQFBgcICQoLEQACAQIEBAMEBwUEBAABAncAAQIDEQQFITEGEkFRB2FxEyIygQgUQpGhscEJIzNS8BVictEKFiQ04SXxFxgZGiYnKCkqNTY3ODk6Q0RFRkdISUpTVFVWV1hZWmNkZWZnaGlqc3R1dnd4eXqCg4SFhoeIiYqSk5SVlpeYmZqio6Slpqeoqaqys7S1tre4ubrCw8TFxsfIycrS09TV1tfY2dri4+Tl5ufo6ery8/T19vf4+fr/2gAMAwEAAhEDEQA/APYqKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAoooo ...193998 bytes
trce: Lagrange.Core.BotContext[0]
      [2024-08-24 06:17:25] [ServiceContext] [DEBUG]: Outgoing SSOFrame: OidbSvcTrpcTcp.0x11c4_100
trce: Lagrange.Core.BotContext[0]
      [2024-08-24 06:17:25] [ServiceContext] [DEBUG]: Incoming SSOFrame: OidbSvcTrpcTcp.0x11c4_100
trce: Lagrange.Core.BotContext[0]
      [2024-08-24 06:17:25] [ServiceContext] [DEBUG]: Outgoing SSOFrame: MessageSvc.PbSendMsg
trce: Lagrange.Core.BotContext[0]
      [2024-08-24 06:17:27] [ServiceContext] [DEBUG]: Incoming SSOFrame: MessageSvc.PbSendMsg
trce: Lagrange.OneBot.Core.Network.Service.ForwardWSService[13]
      Send(72a874e5-6ea5-4e38-978e-866ca7222a93) {"status":"ok","retcode":0,"data":{"message_id":291382815},"echo":3}
trce: Lagrange.Core.BotContext[0]
      [2024-08-24 06:17:27] [ServiceContext] [DEBUG]: Incoming SSOFrame: trpc.msg.olpush.OlPushService.MsgPush
trce: Lagrange.Core.BotContext[0]
      [2024-08-24 06:17:27] [ServiceContext] [DEBUG]: Outgoing SSOFrame: OidbSvcTrpcTcp.0x11c4_200
trce: Lagrange.Core.BotContext[0]
      [2024-08-24 06:17:27] [ServiceContext] [DEBUG]: Incoming SSOFrame: OidbSvcTrpcTcp.0x11c4_200
info: Lagrange.Core.BotContext[0]
      [2024-08-24 06:17:27] [MessagingLogic] [VERBOSE]: [MessageChain(***)(***)] [Image: 1272x990] [图片] d4a2bc2969fc7a0f16a6331c7678088d.jpg 146172 http://gchat.qpic.cn/gchatpic_new/***/***-1187407363-D4A2BC2969FC7A0F16A6331C7678088D/0?term=255&is_origin=1 | [Image: 1272x990] [图片] d4a2bc2969fc7a0f16a6331c7678088d.jpg 146172 https://multimedia.nt.qq.com.cn/download?appid=1407&fileid=CgoxNTgxOTM4MDE0EhRT6YQn3zAxh2E1c94SwExMbAqpPRj89Qgg_woo5afnnpKMiANQgL2jAQ&rkey=CAISKKSBekjVG1fMbogrh4KPqfGB2D9u8hX7D147Y1vmd2VkcpzGw-lQw3Y
trce: Lagrange.Core.BotContext[0]
      [2024-08-24 06:17:27] [ServiceContext] [DEBUG]: Outgoing SSOFrame: Heartbeat.Alive
trce: Lagrange.OneBot.Core.Network.Service.ForwardWSService[13]
      Send(72a874e5-6ea5-4e38-978e-866ca7222a93) {"interval":5000,"status":{"app_initialized":true,"app_enabled":true,"app_good":true,"online":true,"good":true},"meta_event_type":"heartbeat","time":1724451447,"self_id":***,"post_type":"meta_event"}
trce: Lagrange.OneBot.Core.Network.Service.ForwardWSService[13]
      Send(72a874e5-6ea5-4e38-978e-866ca7222a93) {"interval":5000,"status":{"app_initialized":true,"app_enabled":true,"app_good":true,"online":true,"good":true},"meta_event_type":"heartbeat","time":1724451452,"self_id":***,"post_type":"meta_event"}

=============================================================================
bot框架的输出

INFO[1591] [bot] 收到群(***)消息 *** : [CQ:at,qq=***,name=@***] #usage rss 
DEBU[1591] [ws] 向服务器发送请求: &{send_group_msg map[group_id:*** message:[CQ:image,file=base64:///9j/2wCEAAoHBwgHBgoICAgLCgoLDhgQDg0NDh0VFhEYIx8lJCIxsf****KKAP//Z]] 3} 
DEBU[1593] [ws] 接收到API调用返回: {"status":"ok","retcode":0,"data":{"message_id":291382815},"echo":3} 
INFO[1593] [api] 发送群消息(***): [CQ:image,file=d4a2bc2969fc7a0f16a6331c7678088d.image] (id=291382815) 
WARN[1664] [ws] Websocket服务器连接断开...
```

### 补充说明（可选）

NTQQ客户端是看不到url错误的消息的，但是可以被lgr或其他lgr接收到。
在合并转发之后会变成一条正确的图片
好像是因为有个CustomFace
image

### LINKED ISSUE #611: [Bug?]: 上报回复消息时含有多余的数据
### 请确保您已阅读以上注意事项，并勾选下方的确认框。

- [X] 我已经仔细阅读上述内容
- [X] 我已经使用 [最新构建](https://github.com/LagrangeDev/Lagrange.Core/actions/) 测试过，问题依旧存在。
- [X] 我已经在 [Issue Tracker](https://github.com/LagrangeDev/Lagrange.Core/issues) 中找过我要提出的问题，没有找到相同问题的ISSUE。
- [X] 我已知晓并同意，此处仅用于汇报程序中存在的问题。若这个 Issue 是关于其他非程序本身问题，则我的 Issue 可能会被无条件自动关闭或/并锁定。（其它疑问请考虑加入TG群询问或在discussions中提问）

### Lagrange项目

OneBot

### 所使用/依赖的Lagrange项目对应的commit

00c8395d953fd0eb88033b09be872bacf233bdc5

### 运行环境

Linux

### 运行架构

x64

### 连接方式

正向 WebSocket

### 重现步骤

使用者回复一条信息
79a5aa71be126fa26b59da08af7ce0a3
Lagrange上报的信息出现重复的at segment和不必要的空text segment，会让匹配前缀的应用需要额外的处理。需要识别多个at字段的应用逻辑则需要针对回复消息类型进行额外的处理。我暂时不清楚这个上报逻辑算不算错误，如果不算的话，我则自己进行适配即可。    这里有个相关的issue，但是目前的新版本依旧有这个问题 #266。
info: Lagrange.Core.BotContext[0]
      [2024-09-27 16:23:54] [MessagingLogic] [VERBOSE]: [MessageChain(233711588)(1043728417)] [Forward] Time: 9/27/2024 4:16:26 PM Sequence: 6842  | [Mention]: @白州梓(721223121) | [Text]:   | [Mention]: @白州梓(721223121) | [Text]:  试试

### 期望的结果是什么？

info: Lagrange.Core.BotContext[0]
      [2024-09-27 16:23:54] [MessagingLogic] [VERBOSE]: [MessageChain(233711588)(1043728417)] [Forward] Time: 9/27/2024 4:16:26 PM Sequence: 6842  | [Mention]: @白州梓(721223121) | [Text]:  试试

### 实际的结果是什么？

info: Lagrange.Core.BotContext[0]
      [2024-09-27 16:23:54] [MessagingLogic] [VERBOSE]: [MessageChain(233711588)(1043728417)] [Forward] Time: 9/27/2024 4:16:26 PM Sequence: 6842  | [Mention]: @白州梓(721223121) | [Text]:   | [Mention]: @白州梓(721223121) | [Text]:  试试

### 简单的复现代码/链接（可选）

_No response_

### Trace 级别日志记录（可选）

_No response_

### 补充说明（可选）

_No response_

### PR DESCRIPTION
**Pull request by DarkRRb**
- Move `MessageFilter` to [`MessageingLogic@Incoming`](https://github.com/LagrangeDev/Lagrange.Core/blob/2b83478bce294d641ed6d94d0c2f4cf93401f117/Lagrange.Core/Internal/Context/Logic/Implementation/MessagingLogic.cs#L46)
- Allow add custom rules to remove multiple entities
- `MessageFilter` add `ImageEntity` filter (fix #548), add `ForwardEntity` filter (fix #611)

### CODE REVIEW

**Comment by Linwenxuan04** [comment]
I think list of index should be replaced with foreach iteration again(multiple iteration) because the index would change if you remove previous elements in the List<IMessageEntity>

**Comment by dogdie233** [comment]
https://github.com/LagrangeDev/Lagrange.Core/blob/e13da6093b22512aef7a670532fa41a74de40a95/Lagrange.Core/Message/MessageFilter.cs#L64
应该把indexs从小到大排序之后再从后往前进行RemoveAt

File: Lagrange.Core/Message/Entity/ForwardEntity.cs [hunk 1]
```diff
@@ -79,7 +79,7 @@ IEnumerable<Elem> IMessageEntity.PackElement()
             {
                 Text = new Text
                 {
-                    Str = null,
+                    Str = "Officially, custom names are no longer allowed, but Lagrange still requires this field to resolve the displayed name",
```
Comment (pk5ls20) [inline Lagrange.Core/Message/Entity/ForwardEntity.cs:82]: Try to construct the same Text field as the official client

Comment (DarkRRb) [reply Lagrange.Core/Message/Entity/ForwardEntity.cs:82]: This field is the content field of the Mention, tx has turned off its customization, but lgr detects the Mention requires that this field is not empty, I did not find the reason for this, so I put random characters in this field to ensure that lgr's detection does not make a mistake

**Comment by DarkRRb** [comment]
> I think list of index should be replaced with foreach iteration again(multiple iteration) because the index would change if you remove previous elements in the List

@dogdie233 provides a solution, simply sort all indexes from large to small and then remove them

**Comment by DarkRRb** [comment]
> https://github.com/LagrangeDev/Lagrange.Core/blob/e13da6093b22512aef7a670532fa41a74de40a95/Lagrange.Core/Message/MessageFilter.cs#L63-L67
> 
> 
> 应该把indexs从小到大排序之后，再从后往前进行RemoveAt

或许使用 lwx 的方法可以节省成本

**Comment by DarkRRb** [comment]
> I think list of index should be replaced with foreach iteration again(multiple iteration) because the index would change if you remove previous elements in the List

Resolved

**Comment by DarkRRb** [comment]
> https://github.com/LagrangeDev/Lagrange.Core/blob/e13da6093b22512aef7a670532fa41a74de40a95/Lagrange.Core/Message/MessageFilter.cs#L63-L67
> 
> 
> 应该把indexs从小到大排序之后，再从后往前进行RemoveAt

已解决

File: Lagrange.Core/Message/MessageFilter.cs [hunk 2]
```diff
@@ -10,33 +10,60 @@ internal static class MessageFilter
...
+            for (int i = 0; i < images.Length; i++)
             {
-                foreach (var mention in mentions)
+                var imageOld = images[i];
+                if (!Uri.IsWellFormedUriString(imageOld.ImageUrl, UriKind.RelativeOrAbsolute))
+                {
+                    return (x.IndexOf(imageOld), false);
+                }
+
+                var uri = new Uri(imageOld.ImageUrl);
```
Comment (pk5ls20) [inline Lagrange.Core/Message/MessageFilter.cs:38]: This filtering rule **filters out all images** for Emoji sent under the official client TIM `v3.5.8.3228`, I'll provide the Protobuf with the original message that can reproduce this

**Review by pk5ls20 — CHANGES_REQUESTED** [review]

**Review by pk5ls20 — APPROVED** [review]

**Pull request merged**